### ISSUE TITLE
OPAM can't find tools

### ISSUE THREAD

**Issue by ncannasse** [ocaml/opam#2896]
When running `opam init` on my Windows install I'm getting the following output:

```
Checking for available remotes: none.
  - you won't be able to use rsync and local repositories unless you install the

    rsync command on your system.
  - you won't be able to use git repositories unless you install the git command
 on
    your system.
  - you won't be able to use mercurial repositories unless you install the hg
    command on your system.
  - you won't be able to use darcs repositories unless you install the darcs
    command on your system.

[WARNING] No external solver found, one of aspucd, packup and mccs is
          recommended (see
          http://opam.ocaml.org/doc/Install.html#ExternalSolvers for details)
[WARNING] Recommended dependencies -- most packages rely on these:
            - make
            - m4
            - cc

[ERROR] Missing dependencies -- the following commands are required for OPAM to
        operate:
          - curl or wget
          - patch
          - tar
          - unzip

[ERROR] Initialisation failed
OpamStd.OpamSys.Exit(66)
```

But I have all these tools correctly installed as part of my cygwin setup. What method does OPAM uses to detect these tools which could make this false positive?

**Comment by AltGr** [comment]
opam uses `/bin/sh -c "command -v COMMAND"` to check the existence of commands, which should work in Cygwin.
@dra27 has been doing a lot of work on porting opam to Windows and may have more hints about what is going on. In any case, we'd be happy to better support your use-cases! Providing inclusive archives for easier distribution of software is also amongst our next goals.

**Comment by dra27** [comment]
Which build of OPAM are you using?

**Comment by ncannasse** [comment]
hi, I'm using OPAM 1.3.0~dev provided by "Opam32" distrib on https://github.com/fdopen/opam-repository-mingw

**Comment by ncannasse** [comment]
`sh` is correctly installed and I can run `curl` etc from either windows prompt or sh prompt

**Comment by ncannasse** [comment]
Ok, I could find that running opam from the windows cmd shell was not working. It seems the tools detection does not work from here.

Also, I couldn't find documentation about an env var to set the OPAM ROOT, is it necessary to always use --root when using a custom repo?

**Comment by dra27** [comment]
Is C:\cygwin\bin (or C:\cygwin64\bin) in your PATH? IIRC, setting OPAMROOT should be fine - is it not working?

**Comment by dra27** [comment]
You don't have a Cygwin version of opam installed as well, do you? Looking again, I'm surprised by the warning about cc and m4, but perhaps I removed those after Andreas compiled it.

The tools detection (and process invocation) has all changed in OPAM 2, but that's not a lot of help to you this very second!

**Comment by ncannasse** [comment]
@dra27 yes I have cygwin/bin in my PATH (it's in program files x86/cygwin/bin).
Thanks for OPAMROOT tip, I didn't find much documentation about it last time I look

`where opam` only gives me one result so there don't seem to have several installed versions nor I remember installing cygwin version directly.

FYI it works well if I run opam through cygwin shell or sh, it only fails when I run it through windows shell

**Comment by dra27** [comment]
Ah, that's potentially interesting having Cygwin in a non-default location (especially one with spaces) - I'll keep in mind to see what happens when I complete my rebase/sync hopefully during next week. Would you be happy to try a replacement binary then?

**Comment by ncannasse** [comment]
Sure

**Comment by ncannasse** [comment]
Another question while we're at it: it seems that relocating .opam to another directory (and setting OPAMROOT) fails to correctly install the packages. In my case I'm getting the following:

```
$ opam install --root d:/projects/ocamhaxe/.opam sedlex
The following actions will be performed:
  ∗  install gen       0.4                                    [required by sedlex]
  ∗  install ppx_tools 5.0                                    [required by sedlex]
  ∗  install sedlex    1.99.3
===== ∗  3 =====
Do you want to continue ? [Y/n] Y
The following variables are set in your environment, it is advised to unset them for OPAM to work correctly.
 - OCAMLFIND_CONF
Do you want to continue ? [Y/n] Y

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[gen: ./configure] Command started
[ppx_tools: make all] Command started
[ERROR] The compilation of gen failed at "./configure --disable-docs
        --disable-tests --disable-bench".
[gen: ocamlfind remove] Command started
[ERROR] The compilation of ppx_tools failed at "make all".
[ppx_tools: ocamlfind remove] Command started

#=== ERROR while compiling ppx_tools.5.0 ======================================#
# opam-version         1.3.0~dev (fb90257639e7abafad504a79cd571a2f838e0a07)
# os                   win32
# command              make all
# path                 D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/ppx_tools.5.0
# exit-code            2
# env-file             D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/ppx_tools.5.0\ppx_tools-168-42bb52.env
# stdout-file          D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/ppx_tools.5.0\ppx_tools-168-42bb52.out
# stderr-file          D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/ppx_tools.5.0\ppx_tools-168-42bb52.err
### stderr ###
# make: *** Aucune règle pour fabriquer la cible « all ». Arrêt.


#=== ERROR while compiling gen.0.4 ============================================#
# opam-version         1.3.0~dev (fb90257639e7abafad504a79cd571a2f838e0a07)
# os                   win32
# command              ./configure --disable-docs --disable-tests --disable-bench
# path                 D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/gen.0.4
# exit-code            127
# env-file             D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/gen.0.4\gen-168-ccabea.env
# stdout-file          D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/gen.0.4\gen-168-ccabea.out
# stderr-file          D:/projects/ocamhaxe/.opam/4.04.0+mingw32c/build/gen.0.4\gen-168-ccabea.err
### stderr ###
# dash: 1: ./configure: not found



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install sedlex 1.99.3
The following actions failed
  λ  build gen       0.4
  λ  build ppx_tools 5.0
No changes have been performed (edited)
```` 

What are the other files that needs to be changed?

We are trying to distribute an opam + ocaml distribution for Haxe where everything is already installed and setup, with a small config script to setup the env vars and config files.

**Comment by AFodil** [comment]
I have similar issues when I try to install ocamlfind with opam through cygwin
```bash
### stderr ###
# + echo
# /bin/bash: m4 : commande introuvable



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  - install ocamlfind 1.7.2
The following actions failed
  - build conf-m4 1
No changes have been performed

=-=- conf-m4.1 troubleshooting =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
=> This package relies on external (system) dependencies that may be missing.
   `opam depext conf-m4.1' may help you find the correct installation for your
   system.
```

**Comment by kit-ty-kate** [comment]
This should have been fixed in opam 2.2.0 which added native windows support. If this is not the case please do reopen an issue
